File: packages/core/src/tools/fixtures/python/sample.py
# encoding: utf-8
# Copyright 2026 Google LLC
# SPDX-License-Identifier: Apache-2.0

"""
Sample Python file for LSP integration tests.
This file contains various code constructs to test LSP operations.
"""

from typing import List, Optional


class Calculator:
    """A simple calculator class for testing LSP operations."""

    def __init__(self, initial_value: int = 0) -> None:
        self._value = initial_value

    def add(self, n: int) -> None:
        """Add a number to the current value."""
        self._value += n

    def subtract(self, n: int) -> None:
        """Subtract a number from the current value."""
        self._value -= n

    def multiply(self, n: int) -> None:
        """Multiply the current value by a number."""
        self._value *= n

    def divide(self, n: int) -> float:
        """Divide the current value by a number."""
        if n == 0:
            raise ValueError("Division by zero")
        return self._value / n

    def get_value(self) -> int:
        """Get the current value."""
        return self._value

    def reset(self) -> None:
        """Reset the calculator to zero."""
        self._value = 0


class ScientificCalculator(Calculator):
    """An advanced calculator with memory functionality."""

    def __init__(self, initial_value: int = 0) -> None:
        super().__init__(initial_value)
        self._memory = 0

    def store_memory(self) -> None:
        """Store current value in memory."""
        self._memory = self.get_value()

    def recall_memory(self) -> int:
        """Recall value from memory."""
        return self._memory

    def clear_memory(self) -> None:
        """Clear the memory."""
        self._memory = 0

    # Override parent method
    def reset(self) -> None:
        """Reset calculator and memory."""
        super().reset()
        self.clear_memory()

    def compute_square_root(self) -> float:
        """Compute square root of current value."""
        import math
        return math.sqrt(self.get_value())


class AdvancedCalculator(Calculator):
    """A calculator with operation history."""

    def __init__(self, initial_value: int = 0) -> None:
        super().__init__(initial_value)
        self._history: List[str] = []

    def _log_operation(self, operation: str) -> None:
        """Log an operation to history."""
        self._history.append(operation)

    def add(self, n: int) -> None:
        """Add and log the operation."""
        super().add(n)
        self._log_operation(f"add {n}")

    def subtract(self, n: int) -> None:
        """Subtract and log the operation."""
        super().subtract(n)
        self._log_operation(f"subtract {n}")

    def get_history(self) -> List[str]:
        """Get the operation history."""
        return self._history.copy()


def calculate_sum(numbers: List[int], initial: int = 0) -> int:
    """Calculate the sum of a list of numbers."""
    calc = Calculator(initial)
    for num in numbers:
        calc.add(num)
    return calc.get_value()


def calculate_product(numbers: List[int], initial: int = 1) -> int:
    """Calculate the product of a list of numbers."""
    calc = Calculator(initial)
    for num in numbers:
        calc.multiply(num)
    return calc.get_value()


def main() -> None:
    """Main function demonstrating calculator usage."""
    calc = Calculator(10)
    calc.add(5)
    calc.multiply(2)

    result = calc.get_value()
    print(f"Result: {result}")

    # Use ScientificCalculator
    sci_calc = ScientificCalculator(100)
    sci_calc.add(50)
    sci_calc.store_memory()

    # Use AdvancedCalculator
    adv_calc = AdvancedCalculator(200)
    adv_calc.add(100)
    print(adv_calc.get_history())


def run_tests() -> None:
    """Run basic tests."""
    test_calc = Calculator()
    test_calc.add(1)
    assert test_calc.get_value() == 1, "Test failed"


if __name__ == "__main__":
    main()
    run_tests()

File: packages/core/src/tools/fixtures/typescript/sample.ts
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

/**
 * Sample TypeScript file for LSP integration tests.
 * This file contains various code constructs to test LSP operations.
 */

// A simple class for testing goToDefinition and references
export class Calculator {
  private value: number = 0;

  constructor(initialValue?: number) {
    this.value = initialValue ?? 0;
  }

  add(n: number): void {
    this.value += n;
  }

  subtract(n: number): void {
    this.value -= n;
  }

  multiply(n: number): void {
    this.value *= n;
  }

  divide(n: number): number {
    if (n === 0) {
      throw new Error('Division by zero');
    }
    return this.value / n;
  }

  getValue(): number {
    return this.value;
  }

  reset(): void {
    this.value = 0;
  }
}

// A class that extends Calculator for testing implementation and inheritance
export class ScientificCalculator extends Calculator {
  private memory: number = 0;

  storeMemory(): void {
    this.memory = this.getValue();
  }

  recallMemory(): number {
    return this.memory;
  }

  clearMemory(): void {
    this.memory = 0;
  }

  // Override parent method
  override reset(): void {
    super.reset();
    this.clearMemory();
  }

  // Abstract-like method for testing implementations
  computeSquareRoot(): number {
    return Math.sqrt(this.getValue());
  }
}

// Another implementation of Calculator pattern
export class AdvancedCalculator extends Calculator {
  private history: string[] = [];

  private logOperation(operation: string): void {
    this.history.push(operation);
  }

  override add(n: number): void {
    super.add(n);
    this.logOperation(`add ${n}`);
  }

  override subtract(n: number): void {
    super.subtract(n);
    this.logOperation(`subtract ${n}`);
  }

  getHistory(): string[] {
    return [...this.history];
  }
}

// Interface for testing implementation lookup
export interface Computable {
  compute(): number;
  reset(): void;
}

// A function that uses Calculator
export function calculateSum(numbers: number[], initial?: number): number {
  const calc = new Calculator(initial);
  for (const num of numbers) {
    calc.add(num);
  }
  return calc.getValue();
}

// Another function for testing references
export function calculateProduct(numbers: number[], initial?: number): number {
  const calc = new Calculator(initial ?? 1);
  for (const num of numbers) {
    calc.multiply(num);
  }
  return calc.getValue();
}

// Main function demonstrating usage
export function main(): void {
  const calc = new Calculator(10);
  calc.add(5);
  calc.multiply(2);

  // Use ScientificCalculator
  const sciCalc = new ScientificCalculator(100);
  sciCalc.add(50);
  sciCalc.storeMemory();

  // Use AdvancedCalculator
  const advCalc = new AdvancedCalculator(200);
  advCalc.add(100);
}

// Test helper function
function runTests(): void {
  const testCalc = new Calculator();
  testCalc.add(1);
}

// Run main if executed directly
if (require.main === module) {
  main();
  runTests();
}
